aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h

```diff
@@ -241,52 +266,79 @@ void _validate_compressed_sparse_indices_kernel(
 
     const auto batch_dims = cidx.sizes().slice(0, cidx.dim() - 1);
     const auto batch_count = indexCount(batch_dims);
-    const auto batch_idx = at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+    const auto batch_idx =
+        at::arange(batch_count, cidx.options()).view(batch_dims).unsqueeze_(-1);
+
+    const Tensor batch_shape = at::tensor(batch_dims, idx.options());
+    const Tensor batch_idx_stride =
+        at::tensor(idx.strides().slice(0, idx.dim() - 1), idx.options());
+    const Tensor all_idx_offsets = at::arange(
+        batch_idx_stride.mul(batch_shape.sub(1)).sum().add(1).item(),
+        idx.options());
+    const Tensor idx_offsets =
+        all_idx_offsets.reshape({-1, 1})
+            .div(batch_idx_stride, "floor") // TODO: support zero strides
+            .remainder(batch_shape)
+            .mul(batch_idx_stride)
+            .sum(1)
+            .eq(all_idx_offsets)
+            .nonzero()
+            .to(idx.dtype())
+            .view(batch_dims)
+            .unsqueeze_(-1);
```

Comment:
These ops could be done on the CPU, and then moved to the GPU if needed.

Reply:
Maybe, if that is of use, there is a function `make_offset_calculator` that allows to map a linear index to a multi-dim offset.

Reply:
The computation of idx offsets is moved to the kernel.